stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -87,10 +87,195 @@ public enum _ExecutionMode : Equatable {
+                      _ status: CTFStatus) {
+    TFE_AddEagerOpToGraph(op, cTraceContext, retvals, retvalCount, status)
+    checkOk(status)
+  }
+
+  // Finalize the trace graph function and execute it, and return the list of
+  // output tensors from the trace execution. These output tensors are owned by
+  // the caller.
+  func finalizeAndExecute(traceeBasicName: String,
+                          inputs: [Tensor<Float>],
+                          outputs: [CTensorHandle]) -> [CTensorHandle] {
+    // We must be in the `notTracing` enum mode.
+    internalConsistencyCheck(_RuntimeConfig.traceState.getFunction == nil)
+
+    // Finish building the trace graph function.
+    let eagerContext = _TFCGetGlobalEagerContext()
+
+    let tracedFunctionName = "\(traceeBasicName)_\(_RuntimeConfig.traceGraphFunctionCounter)"
+    _RuntimeConfig.traceGraphFunctionCounter += 1
+    debugLog("""
+               Finalizing trace graph func \(tracedFunctionName), with \
+               \(inputs.count) tracee inputs, and \
+               \(outputs.count) return values.
+               """)
+    let graphFn = finalize(tracedFunctionName: tracedFunctionName,
+                           outputs: outputs)
+
+    TFE_ContextAddFunction(eagerContext, graphFn, status)
+    checkOk(status)
+    let op: CTFEOp! = TFE_NewOp(eagerContext, tracedFunctionName, status)
+    checkOk(status)
+
+    let deviceName = _ExecutionContext.global.currentDeviceName
+    if let deviceName = deviceName {
+      debugLog("Placing the trace func on device \(deviceName).")
+      TFE_OpSetDevice(op, deviceName, status)
+      checkOk(status)
+    }
+
+    debugLog("Adding \(inputs.count) tracee input tensors.")
+    internalConsistencyCheck(symbolicInputs.count == inputs.count)
+    for input in inputs {
+      _TFCOpAddInputFromTensorHandle(op, input.handle, status)
+      checkOk(status)
+    }
+
+    // Tell TensorFlow to execute the graph function we built, containing
+    // the trace.
+    let maxReturnValueCount = outputs.count
+    debugLog("""
+               Executing trace func \(tracedFunctionName) with up to \
+               \(maxReturnValueCount) return values.
+               """)
+    var returnValues: [CTensorHandle?]
+    returnValues = Array(repeating: nil, count: maxReturnValueCount)
+    var outputReturnValueCount = Int32(maxReturnValueCount)
+    TFE_Execute(op, &returnValues, &outputReturnValueCount, status) 
+    checkOk(status)
+    debugLog("""
+               returnValues.count=\(returnValues.count), \
+               outputReturnValueCount=\(outputReturnValueCount).
+               """)
+    internalConsistencyCheck(outputReturnValueCount <= returnValues.count)
+
+    // Now that all the output elements have been filled in, remove a level of
+    // optional, and also add concrete outputs.
+    var traceGraphOutputs: [CTensorHandle] = []
+    // Points to an element in `returnValues`.
+    var returnValueIdx = 0
+    for output in outputs {
```

Comment:
Use `outputs.enumerated()` to get both an index and a value, instead of counting manually.

Reply:
In the next PR we'll extend the code here that requires manual counting. Left this comment block on the code here.

```
    // We manually increment `returnValueIdx` below instead of using
    // `outputs.enumerated()`, because the logic will be extended in a future PR
    // that requires manual counting.
```